util/topgen/merge.py

```diff
@@ -437,10 +454,26 @@ def simplify_addr(addrs, xbar):
             continue
 
         # TODO: If no other device in current xbar between the gap?
-        if no_device_in_range(xbar, result[-1], e):
-            result[-1]["size_byte"] = "0x{:x}".format(
-                int(e["base_addr"], 0) + int(e["size_byte"], 0) -
-                int(result[-1]["base_addr"], 0))
+        if no_device_in_range(xbar, dev["name"], result[-1], e):
+            # Determine if size can be power of 2 value
+            smallest_addr_gte = get_next_base_addr(e["base_addr"], xbar,
+                                                   dev["name"])
+
+            # Choose next value
+            if smallest_addr_gte == -1:
+                next_value = 0x100000000
```

Comment:
i'm not quite understanding the rationale for this value here if nothing is located.

Reply:
Assume is the Xbar has 32bit address, if there's no device greater than current device end address, it can use any value up to 2^32 -1.

Reply:
I might not fully understand this... does this mean a device can potentially have the address range N to 0xFFFF_FFFF? And it is up to the device to ensure it returns an error for anything in that range it doesn't actually have?

Reply:
the device we talk here is the device connected to another crossbar. And yes, your understand is correct. If the device is the last device in the current crossbar and connected to another crossbar, it could have maximum power of 2 between the N to 0xFFFF_FFFF...

Wait... It cannot have wider than the leading one bit position... I should've revise the code.
For instance, if the base address is `0x4008_0000` and it is the end of the device (xbar: true), the size cannot be greater than `0x8_0000` to minimize the compare logic. If greater than this, it cannot be simple masking but combination of `>= && <=`.

Reply:
sorry i may not be understanding right... let me try to explain what i thought i was reading.
It sounds like if we are looking at a xbar that has say...two devices..0x4000_0000, 0x1000 and then 0x4008000, 0x3000..

since they're not contiguous, we cannot simplify them down.  However, since there is no overlap, does the simply address end up defining a region from 0x4000_0000~0x4000_1000, and 0x4008_0000 to 0xFFFFFFFF?

It's the last part that feels weird to me... I understand the rationale for having power of 2, but wouldn't we want the closest power of 2 greater than the size_byte of the second range instead of all the way to the end of the possible range?   I guess it doesn't make a difference either way...it just feels a little odd that the address space allocation to "default slave" is not super consistent. 

Sorry if I'm completely misunderstanding the intent.

Reply:
> since they're not contiguous, we cannot simplify them down. However, since there is no overlap, does the simply address end up defining a region from 0x4000_0000~0x4000_1000, and 0x4008_0000 to 0xFFFFFFFF?

Nope. `next_value` is to indicate the virtual next device base address is `0x1_0000_0000`. So that the code tries to optimize (find the max power of 2) between 0x4008_0000 and 0x1_0000_0000, unless I coded wrong :) In the example case you mentioned, the result should be 0x8_0000.

Reply:
This is implemented in `find_pow2_size(addr, next_value):`, right? It looks like this function iterates until it found the maximum power of 2 that is smaller than `next_value - base_addr`... so if `next_value` is set to `0x1_0000_0000`, wouldn't that lead to a larger size than `0x8_0000` in @tjaychen's example? I may be reading this wrong tough :)...

Reply:
ah i see what you're saying now @eunchan ...although see another comment below...

Reply:
Revised the find_max_pow2 code.